### PR TITLE
Initial Level 0 support to the develop branch

### PR DESCRIPTION
**Pull request by mxz297**

### CODE REVIEW

File: src/tool/hpcrun/sample-sources/level0.c [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+//******************************************************************************
+
+static void
+METHOD_FN(init)
+{
+    self->state = INIT;
+}
+
+
+static void
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/sample-sources/level0.c:126]: the indenting for this file is inconsistent.

File: src/tool/hpcrun/gpu/level0/level0-handle-map.h [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+ level0_data_node_t* data
+);
+
+level0_data_node_t**
+level0_handle_map_entry_data_get
+(
+ level0_handle_map_entry_t *entry
+);
+
+#endif
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/gpu/level0/level0-handle-map.h:106]: needs a newline at the end.

File: src/tool/hpcrun/gpu/level0/level0-handle-map.c [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+level0_data_node_t**
+level0_handle_map_entry_data_get
+(
+ level0_handle_map_entry_t *entry
+)
+{
+  return &(entry->data);
+}
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/gpu/level0/level0-handle-map.c:182]: needs a newline at the end.

File: src/tool/hpcrun/gpu/level0/level0-event-map.h [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+ level0_data_node_t* new_entry
+);
+
+void
+level0_event_map_delete
+(
+ ze_event_handle_t event_handle
+);
+
+#endif
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/gpu/level0/level0-event-map.h:81]: needs a newline at the end.

File: src/tool/hpcrun/gpu/level0/level0-event-map.c [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+ ze_event_handle_t event_handle
+)
+{
+  spinlock_lock(&event_map_lock);
+
+  uint64_t key = (uint64_t)event_handle;
+  level0_handle_map_delete(&event_map_root, &event_free_list, key);    
+
+  spinlock_unlock(&event_map_lock);
+}
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/gpu/level0/level0-event-map.c:132]: needs a newline at the end

File: src/tool/hpcrun/gpu/level0/level0-data-node.h [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+(
+  level0_data_node_t* node
+);
+
+void
+level0_data_list_free
+(
+ level0_data_node_t* head
+);
+#endif
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/gpu/level0/level0-data-node.h:115]: needs a newline at the end

File: src/tool/hpcrun/gpu/level0/level0-command-process.h [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+void
+level0_command_end
+(
+  level0_data_node_t* command_node,
+  uint64_t start,
+  uint64_t end
+);
+
+#endif
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/gpu/level0/level0-command-process.h:78]: needs a newline at the end

File: src/tool/hpcrun/gpu/level0/level0-command-process.c [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+
+  cstack_ptr_set(&(ga->next), 0);
+  if (gpu_correlation_id_map_lookup(correlation_id) == NULL) {
+    gpu_correlation_id_map_insert(correlation_id, correlation_id);
+  }
+  gpu_activity_process(&gpu_activity);
+  // gpu_activity_process will delete items in gpu-correlation-id-map
+  // We will need to delete items in gpu-host-correlation-map
+  gpu_host_correlation_map_delete(correlation_id);
+}
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/gpu/level0/level0-command-process.c:247]: needs a newline at the end.

File: src/tool/hpcrun/gpu/level0/level0-command-process.c [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+  gpu_op_placeholder_flags_t gpu_op_placeholder_flags = 0;
+  switch (command_node->type) {
+    case LEVEL0_KERNEL: {
+      gpu_op_placeholder_flags_set(&gpu_op_placeholder_flags, gpu_placeholder_type_kernel);
+      break;
+    }
+    case LEVEL0_MEMCPY: {
+      ze_memory_type_t src_type = command_node->details.memcpy.src_type;
+      ze_memory_type_t dst_type = command_node->details.memcpy.dst_type;
+// TODO: Do we need to distinguish copyin and copyout placeholder? We already have host to host, host to device types to distinguish copyin and copyout
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/gpu/level0/level0-command-process.c:185]: all long lines should be appropriately wrapped

File: src/tool/hpcrun/gpu/level0/level0-command-process.c [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+static void
+level0_memcpy_translate
+(
+  gpu_activity_t* ga,
+  level0_data_node_t* c,
+  uint64_t start,
+  uint64_t end
+)
+{
+  PRINT("level0_memcpy_translate: src_type %d, dst_type %d, size %u\n", c->details.memcpy.src_type, c->details.memcpy.dst_type, c->details.memcpy.copy_size);
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/gpu/level0/level0-command-process.c:101]: this should be wrapped better

File: src/tool/hpcrun/gpu/level0/level0-command-list-map.h [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+level0_commandlist_append_memcpy
+(
+ level0_data_node_t** command_list,
+ ze_memory_type_t src_type,
+ ze_memory_type_t dst_type,
+ size_t copy_size,
+ ze_event_handle_t event,
+ ze_event_pool_handle_t event_pool
+);
+#endif
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/gpu/level0/level0-command-list-map.h:123]: needs a newline at the end.

File: src/tool/hpcrun/gpu/level0/level0-command-list-map.c [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+ ze_memory_type_t dst_type,
+ size_t copy_size,
+ ze_event_handle_t event,
+ ze_event_pool_handle_t event_pool
+)
+{
+  level0_data_node_t* list_entry = level0_commandlist_alloc_memcpy(src_type, dst_type, copy_size, event, event_pool);
+  link_node(command_list, list_entry);  
+  return list_entry;
+}
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/gpu/level0/level0-command-list-map.c:215]: needs a newline at the end

File: src/tool/hpcrun/sample-sources/level0.h [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+typedef struct gpu_activity_t gpu_activity_t;
+typedef struct cct_node_t cct_node_t;
+
+void level0_init();
+void level0_fini();
+int level0_bind();
+
+
+#endif //HPCTOOLKIT_AMD_H
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/sample-sources/level0.h:56]: comment needs adjustment

**Comment by mxz297** [comment]
@jmellorcrummey Changes are made accordingly.

File: src/tool/hpcrun/gpu/level0/level0-command-process.c (hunk 8 above)
Comment (jmellorcrummey) [reply src/tool/hpcrun/gpu/level0/level0-command-process.c:253]: This file still needs a newline at the end.

**Pull request merged**